src/main/java/com/google/javascript/clutz/ImportRenameMapBuilder.java

```diff
@@ -0,0 +1,158 @@
+        && rightHandSide.getFirstChild().matchesQualifiedName("goog.require");
+  }
+
+  private String getGoogModuleId(Node astRoot) {
+    if (astRoot.getFirstChild() == null || !astRoot.getFirstChild().isModuleBody()) {
+      return null;
+    }
+
+    Node moduleBody = astRoot.getFirstChild();
+    for (Node statement : moduleBody.children()) {
+      if (isGoogModuleCall(statement)) {
+        return statement.getFirstChild().getChildAtIndex(1).getString();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Does the actual work of walking over the AST, finding any goog.require() assignments or
+   * destructuring assignments, parsing them, and generating the mappings from local symbol names to
+   * exported symbol names.
+   */
+  private Map<String, String> buildImportRenameMap(String localModuleId, Node astRoot) {
+    if (astRoot.getFirstChild() == null || !astRoot.getFirstChild().isModuleBody()) {
+      return null;
+    }
+
+    Map<String, String> importRenameMap = new HashMap<>();
+    Node moduleBody = astRoot.getFirstChild();
+    for (Node statement : moduleBody.children()) {
+      // `const C = goog.require()`
+      if (isGoogRequireAssignment(statement)) {
+        String importedModuleId = statement.getFirstFirstChild().getChildAtIndex(1).getString();
+        String variableName = statement.getFirstChild().getString();
+
+        String localSymbolName = buildLocalSymbolName(localModuleId, variableName);
+        String exportedSymbolName = buildExportedSymbolName(importedModuleId);
+        importRenameMap.put(localSymbolName, exportedSymbolName);
+        // `const {C, Clazz: RenamedClazz} = goog.require()`
+      } else if (isGoogRequireDestructuringAssignment(statement)) {
+        String importedModuleId =
+            statement.getFirstChild().getChildAtIndex(1).getChildAtIndex(1).getString();
+        for (Node destructured : statement.getFirstFirstChild().children()) {
+          String originalName = destructured.getString();
+          // Destructuring can use the original name `const {A} = goog.require("foo.a")` or rename
+          // it `const {A: RenamedA} = ...`, and closure uses whichever in the symbol name it
+          // generates, so we have to extract it.
+          String variableName;
+          if (destructured.getFirstChild() != null) {
+            variableName = destructured.getFirstChild().getString();
+          } else {
+            variableName = originalName;
+          }
+
+          String localSymbolName = buildLocalSymbolName(localModuleId, variableName);
+          String exportedSymbolName = buildExportedSymbolName(importedModuleId, originalName);
+          importRenameMap.put(localSymbolName, exportedSymbolName);
+        }
+      }
+    }
+
+    return importRenameMap;
+  }
+
+  /**
+   * The exported symbol can take 2 forms - one where it refers to everything that the module
+   * exports and another where it refers to just one thing the module exports. If there's a normal
```

Comment:
it's not clear what a normal import here means. I think what you're trying to explain is the difference between an `exports = ...` vs `exports.foo = ...` export, right? If so, the terminology is named exports for the latter. For the former, there's no good single word as it doesn't have a match in ES6 modules.

Reply:
Rewrote it to reference the actual styles.